pkg-manifest/exportable-manifest/src/index.ts

```diff
@@ -159,19 +159,25 @@ async function replaceWorkspaceProtocolPeerDependency (depName: string, depSpec:
     return depSpec
   }
 
-  // Dependencies with bare "*", "^", "~",">=",">","<=",< versions
-  const workspaceSemverRegex = /workspace:([\^~*]|>=|>|<=|<)/
+  // Dependencies with bare "*", "^", "~",">=",">","<=", "<", version
+  const workspaceSemverRegex = /workspace:([\^~*]|>=|>|<=|<)?(\d+\.\d+\.\d+)?/
```

Comment:
There are 3 cases
e.g. `^1`, `^1.2`, `^1.2.3`

This regex only match full semver.

Reply:
https://semver.org/

Reply:
> There are various pattern in semver
> 
> e.g. `^1`, `^1.2`, `^1.2.3` or `1.2.3-alpha-123456`
> 
> 
> 
> This regex only match basic pattern of semver.

That's right, I missed that. Thank you for letting me know.

Reply:
These expression, like carrot and tilde, are described in the npm document
https://docs.npmjs.com/cli/v6/using-npm/semver

Reply:
Thank you so much for your feedback. As you said, npm semver has a different case than the format in Semantic Versioning 2.0.0 because it has a range syntax.

- X-Ranges
- Tilde Rnages
- Caret Ranges

I've updated the regular expressions to support them.

Thanks, I'll add you as a co-author.

Reply:
If you remove the workspace protocol from both peer dependencies like dependencies, how about unifying the `replaceWorkspaceProtocol` and `replaceWorkspaceProtocolPeerDependency` methods into just using `replaceWorkspaceProtocol`?

Reply:
> If the workspace protocol needs to be removed from peer dependencies as well as dependencies, how about unifying the `replaceWorkspaceProtocol` and `replaceWorkspaceProtocolPeerDependency` methods into a single `replaceWorkspaceProtocol`?

That's good approach! However, I would like to focus on resolving the issue(#8355) in this PR. How about we work on improvements after this issue is resolved?

Reply:
I think this approach addresses issue #8335. And the current regex still has some unsatisfactory aspects. Therefore, unifying the methods seems like a good solution
<img width="467" alt="image" src="https://github.com/user-attachments/assets/332e540b-66f1-4441-ad26-39b288ea7e2c">

Reply:
> And the current regex still has some unsatisfactory aspects.

`/workspace:([\^~*]|>=|>|<=|<)?((\d+|[xX]|\*)(\.(\d+|[xX]|\*)){0,2})?/` That regular expression is being used by `replaceWorkspaceProtocolPeerDependency` to check the workspace protocol and `replace` the correct version.

Are there any situations where you would use that regex to check for something like `1.0.0-alpha-1234`? I'd like to know more.

Reply:
Below version is common use case. Above regex cannot match to it.
And if "workspace remove" logic is not different for peer deps and regular deps, I think it's better to unify them.
![image](https://github.com/user-attachments/assets/38fe41b5-11fd-432c-a6ae-5d69cfbea966)

Reply:
> Below version(pnpm beta) is common use case. Above regex cannot match to it.

I was wondering if there is a situation where using `replaceWorkspaceProtocolPeerDependency` and not having a workspace protocol, such as `1.0.0-alpha-1234`, would be checked through the regex. 

[`replaceWorkspaceProtocolPeerDependency`](https://github.com/pnpm/pnpm/pull/8383/files#diff-ec47cca038265070142680df68517f92d6d7cddbc6d7ca641fc7ac6ec6e773aaR157)

```typescript
  if (!depSpec.includes('workspace:')) { }
    return depSpec
  }
```

If there is no `workspace` protocol (e.g. `1.0.0-alpha-1234` or `9.5.0-beta.3`) through that branch, return it as is.

If the workspace protocol is used, regular expression checking is done via `regex.exec` (this is different from `regex.test`).
The string `1.0.0-alpha-1234 or 9.5.0-beta.3` cannot occur before the regular expression exec.

```typescript
const workspaceSemverRegex = /workspace:([\^~*]|>=|>|<=|<)?((\d+|[xX]|\*)(\.(\d+|[xX]|\*)){0,2})?/
const versionAliasSpecParts = workspaceSemverRegex.exec(depSpec)
```

If I'm missing something, please let me know.

> And if "workspace remove" logic is not different for peer deps and regular deps, I think it's better to unify them.

I agree with that comment. It would be a better solution to unify them, but I still haven't figured out why `replaceWorkspaceProtocol` and `replaceWorkspaceProtocolPeerDependency` are separated. I want to do this incrementally.

Thanks for your feedback.

Reply:
@zkochan Is there a reason why replaceWorkspaceProtocol and replaceWorkspaceProtocolPeerDependency need to be different?